tests/nc-server/simple_server.c

```diff
@@ -0,0 +1,368 @@
+	size_t shutdown_after;
+	size_t shutdown_current;
+	size_t nconn;
+	size_t nconn_max;
+	void * accept_cookie;
+	struct conn_list_node * conn_cookies;
+	void * caller_cookie;
+	int (* callback_nc_message)(void *, uint8_t *, size_t);
+};
+
+/* Doubly linked list. */
+struct conn_list_node {
+	/* General "dispatch"-level info. */
+	struct conn_list_node * prev;
+	struct conn_list_node * next;
+	struct accept_state * A;
+
+	/* Reading a network message. */
+	int sock_read;
+	uint8_t buf[BUFLEN];
+	void * network_read_cookie;
+};
+
+/* Forward definitions. */
+static int callback_read(void *, ssize_t);
+static int callback_gotconn(void *, int);
+static int conndied(struct conn_list_node *);
+static int doaccept(struct accept_state *);
+static int drop(struct conn_list_node *);
+static void simple_server_shutdown(void * cookie);
+
+/* Non-blocking accept, if we can have more connections. */
+static int
+doaccept(struct accept_state * A)
+{
+	int rc = 0;
+
+	/* If we can, accept a new connection. */
+	if ((A->nconn < A->nconn_max) && (A->accept_cookie == NULL) &&
+	    !A->shutdown_requested) {
+		if ((A->accept_cookie =
+		    network_accept(A->s, callback_gotconn, A)) == NULL) {
+			warnp("network_accept");
+			rc = -1;
+		}
+	}
+
+	/* Return success/fail status. */
+	return (rc);
+}
+
+/* A connection has closed.  Accept more if necessary. */
+static int
+conndied(struct conn_list_node * node_ptr)
+{
+	struct accept_state * A = node_ptr->A;
+
+	/* We should always have a non-empty list of conn_cookies. */
+	assert(A->conn_cookies != NULL);
+
+	/* We've lost a connection. */
+	A->nconn -= 1;
+
+	/* Adjust shutdown counter, if relevant. */
+	if (A->shutdown_after > 0) {
+		A->shutdown_current++;
+		if (A->shutdown_current >= A->shutdown_after)
+			A->shutdown_requested = 1;
+	}
+
+	/* Remove the closed connection from the list of conn_cookies. */
+	if (node_ptr == A->conn_cookies) {
+		/* Closed conn_cookie is first in the list. */
+		A->conn_cookies = node_ptr->next;
+		if (node_ptr->next != NULL)
+			node_ptr->next->prev = NULL;
+	} else {
+		/* Closed conn_cookie is in the middle of list. */
+		assert(node_ptr->prev != NULL);
+		node_ptr->prev->next = node_ptr->next;
+		if (node_ptr->next != NULL)
+			node_ptr->next->prev = node_ptr->prev;
+	}
+
+	/* Clean up the now-unused node. */
+	free(node_ptr);
+
+	/* If requested to do so, indicate that all connections are closed. */
+	if (A->shutdown_requested && (A->nconn == 0))
+		A->conndone = 1;
+
+	/* Maybe accept more connections. */
+	return (doaccept(A));
+}
+
+/* Handle an incoming connection. */
+static int
+callback_gotconn(void * cookie, int s)
+{
+	struct accept_state * A = cookie;
+	struct conn_list_node * node_new;
+
+	/* This accept is no longer in progress. */
+	A->accept_cookie = NULL;
+
+	/* If we got a -1 descriptor, something went seriously wrong. */
+	if (s == -1) {
+		warnp("network_accept");
+		goto err0;
+	}
+
+	/* We have gained a connection. */
+	A->nconn += 1;
```

Comment:
Maybe this should be rolled back in the error path if we fail to allocate a node or start reading from the socket?